Adjust SCT runner lifespans based on test failure and collect_logs status

KnifeyMoloko opened the pull request:
This PR aims to modify the way we clean SCT runner instances. We clean SCT runners close to the end of most of our pipelines. The requested logic was to:

- extend the runners `keep` tag to `now()`  +6 hours if the test failed and logs were collected
- allow runners to be removed at the end of the test on demand --> I set them to be removed by default if the test passed with "SUCCESS" and logs were collected

As an additional goal this PR removes logic from `list_sct_runners`, since part of the logic for keeping or cleaning the runner was placed there, modifying the listed representations of instances. IMHO a `list_*` method should not modify these representations at all.

Trello tasks: https://trello.com/c/2ypkHwF7, https://trello.com/c/6Lr0Rowm
GCE run with a failure before the test duration (extend the SCT keep time): https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/delete-sct-runners-by-request/106
AWS run with a failure before the test duration (extend the SCT keep time): https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/delete-sct-runners-by-request/107

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

fgelcer commented on sct.py:
```diff
@@ -1372,13 +1373,24 @@ def create_runner_instance(cloud_provider, region, availability_zone, instance_t
 @click.option("-ip", "--runner-ip", required=False, type=str, default="")
 @click.option('--dry-run', is_flag=True, default=False, help='dry run')
-def clean_runner_instances(test_status, runner_ip, dry_run):
+@click.option("--logs-collected", default=False, help='Were the logs collected')
```
in commands bash like, we usually don't send `True` or `False`, but add the parameter (or not)...
so it would most likely have `is_flag=True` in it, so no values must be passed in addition to the parameter itself..
so the code that will make this call in the `groovy` file, will check `if logs collected; then cmd += --logs-collected`

KnifeyMoloko replied:
@fgelcer I changed the name of this argument to "collect_logs_result", since that's the name of the stage that we're interested in here. I'm hesitant to treat this as a flag, since it's a piece of data we're passing in to the function, not a switch for the function's behavior (in contrast to `--dry-run` here.

@fruch It is a boolean, but as mentioned before, it's not something that we enable / disable, it's data about a different stage's success or failure that we're passing in here. I can add an explanation for how this data is treated, but it would quite verbose for a CLI help text snippet, since it will modify the SCT runners tags based on logic involving also checking the runner's log etc.

fgelcer replied:
maybe it would be clearer if it was called `were_logs_collected` or something like that?

fruch commented on sdcm/sct_runner.py:
```diff
@@ -256,6 +257,12 @@ def _create_instance(self,
+    def set_tags(self,
+                 sct_runner_info: SctRunnerInfo,
+                 tags: dict):
+        ...
```
why this implementation is empty ?, all the implementations below look identical...
or it's not yet tested on the 3 providers ?

KnifeyMoloko replied:
This was likely the abstract method or just a placeholder. Either way the method has implementations in each concrete child class now.

fruch commented on vars/cleanSctRunners.groovy:
```diff
@@ -13,11 +15,19 @@ def call(Map params, RunWrapper currentBuild){
     if [[ "$cloud_provider" == "aws" || "$cloud_provider" == "gce" || "$cloud_provider" == "azure" ]]; then
         export RUNNER_IP=\$(cat sct_runner_ip||echo "")
-        ./docker/env/hydra.sh clean-runner-instances --test-status "$test_status" --runner-ip \${RUNNER_IP}
+        ./docker/env/hydra.sh clean-runner-instances --test-status "$test_status" --runner-ip \${RUNNER_IP} --logs-collected "$collect_logs_status"
```
what's `--logs-collected` mean in context of cleaning instances ?, it's not exactly clear

KnifeyMoloko replied:
As explained above: it's used by the cleanup logic for resetting the tag values based on if the test failed before it's target duration and if the logs were collected in an earlier stage.

fruch commented:
@KnifeyMoloko FYI https://github.com/scylladb/scylla-cluster-tests/issues/5051, maybe you should also remove the network tag `keep-alive` from GCE runners.

KnifeyMoloko commented on sct.py:
```diff
@@ -1372,13 +1373,24 @@ def create_runner_instance(cloud_provider, region, availability_zone, instance_t
+@click.option("-t", "--tags", type=(str, str),
+              help="Space separated key value pair to add as a new tag to the runner",
+              multiple=True)
+def set_runner_tags(runner_ip, tags):
```
I agree, though I'd prefer to do that in a different PR, this one is large enough as it is.

fgelcer replied:
@KnifeyMoloko , i agree, so please open a task to not miss it

roydahan replied:
low priority task

KnifeyMoloko replied:
Task created in the NEW queue: https://trello.com/c/6kPMUvJL/5335-create-a-sct-cli-method-for-tagging-retagging-instances

KnifeyMoloko commented:
> @KnifeyMoloko FYI #5051, maybe you should also remove the network tag `keep-alive` from GCE runners.

This PR will change the labels on our GCE instances, so it should help with that, but I need to take a closer look at #5051 too.

soyacz commented:
Generally I dislike changes in `clean_sct_runners` - as a user of api, I don't know if logs were collected or not. So I don't know what to pass to `--collect_logs_result` argument. 
I also dislike idea of passing this param here and there - it's very confusing and complicated.
Why cannot it be done with simple solution like:
1. Start the test, set sct-runner tag to keep:alive
2. After test finish, `collect-logs` step after successful collection sets `keep: terminate` tag
3. Clean sct-runner function just cleans in standard way
Another solution: instead of changing `keep` tag, collect-logs function could add tag `logs_collected: true' and later, `clean_sct_runners` function could take this tag into cleaning logic.

KnifeyMoloko commented:
> Generally I dislike changes in `clean_sct_runners` - as a user of api, I don't know if logs were collected or not. So I don't know what to pass to `--collect_logs_result` argument. I also dislike idea of passing this param here and there - it's very confusing and complicated. Why cannot it be done with simple solution like:
> 
> 1. Start the test, set sct-runner tag to keep:alive
> 2. After test finish, `collect-logs` step after successful collection sets `keep: terminate` tag
> 3. Clean sct-runner function just cleans in standard way
>    Another solution: instead of changing `keep` tag, collect-logs function could add tag `logs_collected: true` and later, `clean_sct_runners` function could take this tag into cleaning logic.

@soyacz The main users of the API are our pipeline scripts and we're already passing other test-related information there (or we're making some hardcoded assumptions about that information), e.g. test status. That being said I do think the idea of `collect_logs` adding a tag to the runner (it would be more efficient than tagging all the nodes) - I'll give that a try.

roydahan reviewed (changes requested):
My comments here will be general:

First, I think it's an overkill, and I would simply kill the runner.
If i wanted to invest time, I would invest more time on retry the collect logs with larger timeout or something like that.

Since, you already did it, it looks ok to me except:
1. The main commit doesn't have commit message, it should probably include everything you wrote in the PR description.
2. I don't think the separation of commits is correct and I think they should all be squashed into 1 commit. none of them is really independent by its own.

KnifeyMoloko commented:
> My comments here will be general:
> 
> First, I think it's an overkill, and I would simply kill the runner. If i wanted to invest time, I would invest more time on retry the collect logs with larger timeout or something like that.
> 
> Since, you already did it, it looks ok to me except:
> 
> 1. The main commit doesn't have commit message, it should probably include everything you wrote in the PR description.
> 2. I don't think the separation of commits is correct and I think they should all be squashed into 1 commit. none of them is really independent by its own.

Re; 1 - I added a more descriptive commit message
Re: 2 - I fixed up the commits, it's all in 1 commit now

soyacz commented on sdcm/sct_runner.py:
```diff
@@ -930,39 +1036,107 @@ def list_sct_runners(test_status: Optional[str] = None, test_runner_ip: str = No
+        utc_now = datetime.datetime.now(tz=datetime.timezone.utc)
+        LOGGER.info("UTC now: %s", utc_now)
+
+        runners_cleaned = 0
```
I think we don't need it more.

KnifeyMoloko replied:
Good catch. Fixed (removed, edited the final log message)

soyacz reviewed (changes requested):
I don't see where  set `logs_collected` tag - I think it is missing.

fgelcer reviewed (commented):
other than few comments that are still pending, LGTM

KnifeyMoloko commented:
> I don't see where set `logs_collected` tag - I think it is missing.

It got lost with the commit fixup / squashing, I returned it back.

KnifeyMoloko commented on sdcm/sct_runner.py:
```diff
@@ -803,14 +889,16 @@ def _create_instance(self,  # pylint: disable=too-many-arguments
                 "admin_public_key": self.key_pair.public_key.decode(),
             }
-            return azure_region.create_virtual_machine(
+            self.instance = azure_region.create_virtual_machine(
```
Task creatrd in the "NEW" Trello column: https://trello.com/c/la55Gle5

fgelcer reviewed (approved):
LGTM

soyacz commented:
@KnifeyMoloko I've tried running it `./sct.py clean-runner-instances --dry-run` and failed due not finding `eu-north-1a` region in `AwsRegion`.

KnifeyMoloko commented:
Apart from a fix for the region issue @soyacz encountered, a change is needed to accommodate the new vpc peering we have in AWS, since we can't remote into runner instances directly now. I'll add a fix shortly.

soyacz commented on sdcm/sct_runner.py:
```diff
@@ -930,39 +1040,111 @@ def list_sct_runners(test_status: Optional[str] = None, test_runner_ip: str = No
-                continue
+            if "alive" in str(sct_runner_info.keep):
+                LOGGER.info("Skip %s because `keep' == `alive. No runners have been terminated'", sct_runner_info)
+                return
```
when doing `return` we stop after first node. Is it intended?

fgelcer replied:
we should have only a single SCT Runner, no?

KnifeyMoloko replied:
We have a `hydra-clean-runner-instances.jenkinsfile` that uses the `clean-runner-instances` command without providing any runner ips. I'm not sure this is really needed (and it's a bit of playing with fire), but I didn't want to remove that functionality. In most cases though we'll be using this by providing a runner ip and test status, so only one runner will be managed.

As for the `continue` vs `return`, it's fixed.

soyacz reviewed (changes requested):
I executed it again and finished after 1 node (due to `return` - see my comment).

Also it was showing a prompt about not known host key - I believe there's an error in `get_proxy_command` that is not setting `-o StrictHostKeyChecking=no`

fgelcer reviewed (commented):
little comments, but LGTM

soyacz commented on sdcm/sct_runner.py:
```diff
@@ -955,14 +1145,13 @@ def clean_sct_runners(test_status: Optional[str] = None, test_runner_ip: str = N
         if dry_run:
             LOGGER.info("Skip %s because of dry-run", sct_runner_info)
-            continue
+            return
```
should continue here too, so dry run behaves like in normal run

KnifeyMoloko replied:
Fixed

soyacz reviewed (approved):
LGTM

KnifeyMoloko commented:
Waiting for https://github.com/scylladb/scylla-cluster-tests/pull/5121 to be merged for the unit tests to pass

fgelcer reviewed (approved):
LGTM

soyacz reviewed (approved):
LGTM
